source/common/upstream/original_dst_cluster.cc

```diff
@@ -0,0 +1,145 @@
+#include "common/upstream/original_dst_cluster.h"
+
+#include <chrono>
+#include <list>
+#include <string>
+#include <vector>
+
+#include "common/network/address_impl.h"
+#include "common/network/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// Static cast below is guaranteed to succeed, as code instantiating the cluster
+// configuration, that is run prior to this code, checks that an OriginalDstCluster is
+// always configured with an OriginalDstCluster::LoadBalancer, and that an
+// OriginalDstCluster::LoadBalancer is never configured with any other type of cluster,
+// and throws an exception otherwise.
+
+OriginalDstCluster::LoadBalancer::LoadBalancer(HostSet& host_set, ClusterSharedPtr& parent)
+    : host_set_(host_set), parent_(std::static_pointer_cast<OriginalDstCluster>(parent)),
+      info_(parent->info()) {
+  // host_set_ is initially empty.
+  host_set_.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                     const std::vector<HostSharedPtr>& hosts_removed) -> void {
+    // Update the hosts map
+    for (const HostSharedPtr& host : hosts_removed) {
+      ENVOY_LOG(debug, "Removing host {}.", host->address()->asString());
+      host_map_.remove(host);
+    }
+    for (const HostSharedPtr& host : hosts_added) {
+      if (host_map_.insert(host)) {
+        ENVOY_LOG(debug, "Adding host {}.", host->address()->asString());
+      }
+    }
+  });
+}
+
+HostConstSharedPtr
+OriginalDstCluster::LoadBalancer::chooseHost(const LoadBalancerContext* context) {
+  if (context) {
+    const Network::Connection* connection = context->downstreamConnection();
+
+    // The local address of the downstream connection is the original destination address,
+    // if usingOriginalDst() returns 'true'.
+    if (connection && connection->usingOriginalDst()) {
+      const Network::Address::Instance& dst_addr = connection->localAddress();
+
+      // Check if a host with the destination address is already in the host set.
+      HostSharedPtr host = host_map_.find(dst_addr);
+      if (host) {
+        ENVOY_LOG(debug, "Using existing host {}.", host->address()->asString());
+        host->used(true); // Mark as used.
+        return std::move(host);
+      }
+      // Add a new host
+      const Network::Address::Ip* dst_ip = dst_addr.ip();
+      if (dst_ip) {
+        Network::Address::InstanceConstSharedPtr host_ip_port;
+        if (dst_ip->version() == Network::Address::IpVersion::v4) {
+          host_ip_port.reset(
+              new Network::Address::Ipv4Instance(dst_ip->addressAsString(), dst_ip->port()));
+        } else {
+          host_ip_port.reset(
+              new Network::Address::Ipv6Instance(dst_ip->addressAsString(), dst_ip->port()));
+        }
```

Comment:
That's fine. Can you add a TODO for now?

Also, can you update the title and commit message via the UI in this PR so that when we merge this we can just copy them in? Thanks.